<a href="https://colab.research.google.com/github/athulvingt/Coursera_Capstone/blob/main/clustering_neighbourhood.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Segmenting and Clustering Neighborhoods in Toronto
Applied Data Science Capstone Assignment

## Part 1 : Data scraping  
scrape the html page and wrangle Toronto neighbourhood data </br>
source : wikipedia https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M 

In [1]:
# Importing modules

import numpy as np
import pandas as pd
from bs4 import BeautifulSoup
from urllib.request import urlopen

In [2]:
# Function to parse HTML file

def GetHTML(url):
  html = urlopen(url)
  soup = BeautifulSoup(html,'html.parser')
  return soup

In [3]:
# Looking for tables in the webpage

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
content = GetHTML(url)
tables = content.find_all('table')
print(len(tables))

3


In [4]:
for t in tables:
  print(t.prettify)

<bound method Tag.prettify of <table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</t

In [5]:
# There are three tables and we need the content inside the table
# <table class="wikitable sortable"> or the first table (tables[0])

tabel = content.find('table',{'class':'wikitable sortable'})
rows = tabel.find_all('tr')[1:]

In [6]:
# iterating rows and creating a strunctured pandas dataframe

data = []
for row in rows:
  cells = row.find_all('td')
  b = cells[1].get_text().replace("\n","")
  if b in 'Not assigned':
    continue
  p = cells[0].get_text().replace("\n","")
  n = cells[2].get_text().replace("\n","")
  if n in 'Not assigned':
    n = b
  data.append({  
      'PostalCode':p,
      'Borough':b,
      'Neighborhood': n
    })

In [7]:
scrape_data = pd.DataFrame(data)
scrape_data.head(10)

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village
6        M1B       Scarborough                               Malvern, Rouge
7        M3B        North York                                    Don Mills
8        M4B         East York              Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto                     Garden District, Ryerson

In [105]:
scrape_data.describe(include='all')

PostalCode     Borough Neighborhood
count         103         103          103
unique        103          10           99
top           M4B  North York    Downsview
freq            1          24            4

In [106]:
scrape_data.loc[scrape_data['Neighborhood']=='Downsview']

PostalCode     Borough Neighborhood
40        M3K  North York    Downsview
46        M3L  North York    Downsview
53        M3M  North York    Downsview
60        M3N  North York    Downsview

There are 4 neighborhoods with same name  but different Postal Code

In [8]:
scrape_data.to_csv('Toronto_neighborhood.csv',index=False)
scrape_data.shape

(103, 3)

## Part 2 : Geocoding

In [9]:
# Importing modules

import pandas as pd
import numpy as np

#  Using Google maps to geocode the address
!pip install -q googlemaps
from googlemaps import Client as GoogleMaps

In [10]:
data = pd.read_csv('Toronto_neighborhood.csv')
data.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [11]:
def lat_lon(row):
  postal = row[0]
  try:
      geocode_result = gmaps.geocode('{}, Toronto, Ontario'.format(postal))
      lat = geocode_result[0]['geometry']['location'] ['lat']
      lng = geocode_result[0]['geometry']['location']['lng']
  except:
      print(postal)

  return pd.Series([lat, lng])

In [12]:
YOUR_API_KEY = '' # enter your API KEY here


gmaps = GoogleMaps(YOUR_API_KEY)
data[['lat','lon']] = data.apply(lat_lon, axis=1)
data.head()

PostalCode           Borough  ...        lat        lon
0        M3A        North York  ...  43.753259 -79.329656
1        M4A        North York  ...  43.725882 -79.315572
2        M5A  Downtown Toronto  ...  43.654260 -79.360636
3        M6A        North York  ...  43.718518 -79.464763
4        M7A  Downtown Toronto  ...  43.662301 -79.389494

[5 rows x 5 columns]

In [13]:
data.to_csv('Toronto_neighborhood_cord.csv',index=False)

## Part 3 : Neighborhood Clustering

In [14]:
## Importing modules


import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install -q geopy  # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')


Libraries imported.


In [15]:
## Reading the data from csv file

neighborhoods = pd.read_csv('Toronto_neighborhood_cord.csv')
neighborhoods.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

         lat        lon  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [16]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


Cluster only the neighborhoods in North York. So let's slice the original dataframe and create a new dataframe of the North York data.

In [17]:
northyork_data = neighborhoods[neighborhoods['Borough'] == 'North York'].reset_index(drop=True)
northyork_data.head()

PostalCode     Borough                      Neighborhood        lat  \
0        M3A  North York                         Parkwoods  43.753259   
1        M4A  North York                  Victoria Village  43.725882   
2        M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3        M3B  North York                         Don Mills  43.745906   
4        M6B  North York                         Glencairn  43.709577   

         lon  
0 -79.329656  
1 -79.315572  
2 -79.464763  
3 -79.352188  
4 -79.445073

Let's take the geographical coordinates of North York, Toronto 

In [18]:
address = 'North York, Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 43.7543263, -79.44911696639593.


Visualizing North York Neighborhoods on a folium map

In [19]:
# create map of North York using latitude and longitude values
map_northyork = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(northyork_data['lat'], northyork_data['lon'], northyork_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_northyork)  
    
map_northyork

#### Define Foursquare Credentials and Version

In [ ]:
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

### Test Case - 1 (Considereing any one neighborhood)

Expolore one of the neighborhood

In [21]:
northyork_data.loc[0, 'Neighborhood']

'Parkwoods'

Get the neighborhood's latitude and longitude values.

In [22]:
neighborhood_latitude = northyork_data.loc[0, 'lat'] # neighborhood latitude value
neighborhood_longitude = northyork_data.loc[0, 'lon'] # neighborhood longitude value

neighborhood_name = northyork_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

In [23]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API


radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fd783d2a06d855f1b5e63f6'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4e8d9dcdd5fbbbb6b3003c7b-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d163941735',
         'name': 'Park',
         'pluralName': 'Parks',
         'primary': True,
         'shortName': 'Park'}],
       'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'location': {'address': 'Toronto',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'distance': 245,
        'formattedAddress': ['Toronto', 'Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'lat

In [24]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

 Clean the json and structure it into a _pandas_ dataframe.

In [25]:
venues = results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues) # flatten JSON   
nearby_venues.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


referralId  reasons.count  \
0  e-0-4e8d9dcdd5fbbbb6b3003c7b-0              0   
1  e-0-4cb11e2075ebb60cd1c4caad-1              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id       venue.name venue.location.address  \
0  4e8d9dcdd5fbbbb6b3003c7b  Brookbanks Park                Toronto   
1  4cb11e2075ebb60cd1c4caad    Variety Store   29 Valley Woods Road   

   venue.location.lat  venue.location.lng  \
0           43.751976          -79.332140   
1           43.751974          -79.333114   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': 43.75197604605557...                      245   
1  [{'label': 'display', 'lat': 43.75197441585782...                      312   

  venue.location.cc venue.location.city venue.location.state  \
0                CA             Toronto                   ON   
1                CA             Toronto                   ON   

  venue.location.country             venue.location.formattedAddress  \
0                 Canada               [Toronto, Toronto ON, Canada]   
1                 Canada  [29 Valley Woods Road, Toronto ON, Canada]   

                                    venue.categories  venue.photos.count  \
0  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...                   0   
1  [{'id': '4bf58dd8d48988d1f9941735', 'name': 'F...                   0   

  venue.photos.groups venue.venuePage.id  
0                  []          600917367  
1                  []                NaN

In [26]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name                                   venue.categories  \
0  Brookbanks Park  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...   
1    Variety Store  [{'id': '4bf58dd8d48988d1f9941735', 'name': 'F...   

   venue.location.lat  venue.location.lng  
0           43.751976          -79.332140  
1           43.751974          -79.333114

In [27]:
# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

In a 500 meter range we could retrive only 2 venues near to Parkwood.

## Part 3B : Clustering and segmenting all Neighborhoods in Toronto.

Let's create a function to repeat the same process to all the neighborhoods in Toronto

There are 4 rows with same neighborhood in the dataframe, still the postal codes are unique, I think its better to use Postal code as frame of reference

In [110]:
def getNearbyVenues(names, postal, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, post, lat, lng in zip(names, postal, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            post, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [111]:
#get nearby venues

Toronto_data = neighborhoods.copy()
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighborhood'],
                                   postal = Toronto_data['PostalCode'],
                                   latitudes=Toronto_data['lat'],
                                   longitudes=Toronto_data['lon'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [112]:
print(Toronto_venues.shape)
Toronto_venues.head()

(2133, 7)


PostalCode  Neighborhood Latitude  Neighborhood Longitude  \
0        M3A              43.753259              -79.329656   
1        M3A              43.753259              -79.329656   
2        M4A              43.725882              -79.315572   
3        M4A              43.725882              -79.315572   
4        M4A              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

In [113]:
Venue_cat =  Toronto_venues['Venue Category'].unique()
print(sorted(Venue_cat))

['Accessories Store', 'Afghan Restaurant', 'Airport', 'Airport Food Court', 'Airport Gate', 'Airport Lounge', 'Airport Service', 'Airport Terminal', 'American Restaurant', 'Antique Shop', 'Aquarium', 'Art Gallery', 'Art Museum', 'Arts & Crafts Store', 'Asian Restaurant', 'Athletics & Sports', 'Auto Garage', 'Auto Workshop', 'BBQ Joint', 'Baby Store', 'Bagel Shop', 'Bakery', 'Bank', 'Bar', 'Baseball Field', 'Baseball Stadium', 'Basketball Stadium', 'Beach', 'Bed & Breakfast', 'Beer Bar', 'Beer Store', 'Belgian Restaurant', 'Bike Shop', 'Bistro', 'Boat or Ferry', 'Bookstore', 'Boutique', 'Brazilian Restaurant', 'Breakfast Spot', 'Brewery', 'Bridal Shop', 'Bubble Tea Shop', 'Building', 'Burger Joint', 'Burrito Place', 'Bus Line', 'Bus Station', 'Business Service', 'Butcher', 'Café', 'Cajun / Creole Restaurant', 'Camera Store', 'Candy Store', 'Caribbean Restaurant', 'Cheese Shop', 'Chinese Restaurant', 'Chocolate Shop', 'Church', 'Climbing Gym', 'Clothing Store', 'Cocktail Bar', 'Coffee Sh

Everything looks good, But i dont understand why there is a venue category called Neighborhood.

**Lets explore the venue category Neighborhood**

In [114]:
Toronto_venues.loc[Toronto_venues['Venue Category'] == 'Neighborhood']

PostalCode  Neighborhood Latitude  Neighborhood Longitude  \
358         M4E              43.676357              -79.293031   
605         M5H              43.650571              -79.384568   
793         M5J              43.640816              -79.381752   
1269        M4M              43.659526              -79.340923   

                 Venue  Venue Latitude  Venue Longitude Venue Category  
358      Upper Beaches       43.680563       -79.292869   Neighborhood  
605   Downtown Toronto       43.653232       -79.385296   Neighborhood  
793       Harbourfront       43.639526       -79.380688   Neighborhood  
1269       Leslieville       43.662070       -79.337856   Neighborhood

I looked these places ( Upper Beaches, Downtown Toronto, Harbourfront	, Leslieville	) on internet and it seems they are just neighborhoods and not venues

So I am **removing these rows from Toronto_venues dataframe**

In [115]:
Toronto_venues.drop(Toronto_venues.index[Toronto_venues['Venue Category'] == 'Neighborhood'], inplace = True)

In [116]:
Toronto_venues.groupby('PostalCode').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
PostalCode                                                         
M1B                             1                       1      1   
M1C                             2                       2      2   
M1E                             8                       8      8   
M1G                             4                       4      4   
M1H                             8                       8      8   
M1J                             3                       3      3   
M1K                             4                       4      4   
M1L                            10                      10     10   
M1M                             2                       2      2   
M1N                             4                       4      4   
M1P                             5                       5      5   
M1R                             6                       6      6   
M1S                             5                       5      5   
M1T                            12                      12     12   
M1V                             4                       4      4   
M1W                            12                      12     12   
M2H                             5                       5      5   
M2J                            70                      70     70   
M2K                             4                       4      4   
M2L                             1                       1      1   
M2N                            33                      33     33   
M2P                             2                       2      2   
M2R                             5                       5      5   
M3A                             2                       2      2   
M3B                             4                       4      4   
M3C                            19                      19     19   
M3H                            21                      21     21   
M3J                             7                       7      7   
M3K                             2                       2      2   
M3L                             6                       6      6   
M3M                             4                       4      4   
M3N                             4                       4      4   
M4A                             6                       6      6   
M4B                            11                      11     11   
M4C                             8                       8      8   
M4E                             3                       3      3   
M4G                            33                      33     33   
M4H                            24                      24     24   
M4J                             3                       3      3   
M4K                            43                      43     43   
M4L                            19                      19     19   
M4M                            36                      36     36   
M4N                             3                       3      3   
M4P                             9                       9      9   
M4R                            18                      18     18   
M4S                            33                      33     33   
M4T                             2                       2      2   
M4V                            14                      14     14   
M4W                             4                       4      4   
M4X                            48                      48     48   
M4Y                            75                      75     75   
M5A                            44                      44     44   
M5B                           100                     100    100   
M5C                            85                      85     85   
M5E                            55                      55     55   
M5G                            68                      68     68   
M5H                            99                      99     99   


In [117]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 272 uniques categories.


In [118]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['PostalCode'] = Toronto_venues['PostalCode'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()


PostalCode  Accessories Store  Afghan Restaurant  Airport  \
0        M3A                  0                  0        0   
1        M3A                  0                  0        0   
2        M4A                  0                  0        0   
3        M4A                  0                  0        0   
4        M4A                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0     0    0               0                 0                   0      0   
1     0    0               0                 0                   0      0   
2     0    0               0                 0                   0      0   
3     0    0               0                 0                   0      0   
4     0    0               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Pl

In [119]:
Toronto_onehot.shape

(2129, 273)

 #### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [120]:
Toronto_grouped = Toronto_onehot.groupby('PostalCode').mean().reset_index()
Toronto_grouped

PostalCode  Accessories Store  Afghan Restaurant  Airport  \
0         M1B           0.000000           0.000000   0.0000   
1         M1C           0.000000           0.000000   0.0000   
2         M1E           0.000000           0.000000   0.0000   
3         M1G           0.000000           0.000000   0.0000   
4         M1H           0.000000           0.000000   0.0000   
5         M1J           0.000000           0.000000   0.0000   
6         M1K           0.000000           0.000000   0.0000   
7         M1L           0.000000           0.000000   0.0000   
8         M1M           0.000000           0.000000   0.0000   
9         M1N           0.000000           0.000000   0.0000   
10        M1P           0.000000           0.000000   0.0000   
11        M1R           0.166667           0.000000   0.0000   
12        M1S           0.000000           0.000000   0.0000   
13        M1T           0.000000           0.000000   0.0000   
14        M1V           0.000000           0.000000   0.0000   
15        M1W           0.000000           0.000000   0.0000   
16        M2H           0.000000           0.000000   0.0000   
17        M2J           0.000000           0.000000   0.0000   
18        M2K           0.000000           0.000000   0.0000   
19        M2L           0.000000           0.000000   0.0000   
20        M2N           0.000000           0.000000   0.0000   
21        M2P           0.000000           0.000000   0.0000   
22        M2R           0.000000           0.000000   0.0000   
23        M3A           0.000000           0.000000   0.0000   
24        M3B           0.000000           0.000000   0.0000   
25        M3C           0.000000           0.000000   0.0000   
26        M3H           0.000000           0.000000   0.0000   
27        M3J           0.000000           0.000000   0.0000   
28        M3K           0.000000           0.000000   0.5000   
29        M3L           0.000000           0.000000   0.0000   
30        M3M           0.000000           0.000000   0.0000   
31        M3N           0.000000           0.000000   0.0000   
32        M4A           0.000000           0.000000   0.0000   
33        M4B           0.000000           0.000000   0.0000   
34        M4C           0.000000           0.000000   0.0000   
35        M4E           0.000000           0.000000   0.0000   
36        M4G           0.000000           0.000000   0.0000   
37        M4H           0.000000           0.000000   0.0000   
38        M4J           0.000000           0.000000   0.0000   
39        M4K           0.000000           0.000000   0.0000   
40        M4L           0.000000           0.000000   0.0000   
41        M4M           0.000000           0.000000   0.0000   
42        M4N           0.000000           0.000000   0.0000   
43        M4P           0.000000           0.000000   0.0000   
44        M4R           0.000000           0.000000   0.0000   
45        M4S           0.000000           0.000000   0.0000   
46        M4T           0.000000           0.000000   0.0000   
47        M4V           0.000000           0.000000   0.0000   
48        M4W           0.000000           0.000000   0.0000   
49        M4X           0.000000           0.000000   0.0000   
50        M4Y           0.000000           0.013333   0.0000   
51        M5A           0.000000           0.000000   0.0000   
52        M5B           0.000000           0.000000   0.0000   
53        M5C           0.000000           0.000000   0.0000   
54        M5E           0.000000           0.000000   0.0000   
55        M5G           0.000000           0.000000   0.0000   
56        M5H           0.000000           0.000000   0.0000   
57        M5J           0.000000           0.000000   0.0000   
58        M5K           0.000000           0.000000   0.0000   
59        M5L           0.000000           0.000000   0.0000   
60        M5M           0.000000           0.000000   0.0000   
61        M5N           0.000000   

In [121]:
Toronto_grouped.shape

(100, 273)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [122]:
num_top_venues = 5

for hood in Toronto_grouped['PostalCode']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['PostalCode'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M1B----
                             venue  freq
0             Fast Food Restaurant   1.0
1                Accessories Store   0.0
2               Mexican Restaurant   0.0
3              Monument / Landmark   0.0
4  Molecular Gastronomy Restaurant   0.0


----M1C----
                        venue  freq
0  Construction & Landscaping   0.5
1                         Bar   0.5
2           Accessories Store   0.0
3   Middle Eastern Restaurant   0.0
4         Monument / Landmark   0.0


----M1E----
                venue  freq
0   Electronics Store  0.12
1      Breakfast Spot  0.12
2      Medical Center  0.12
3                Bank  0.12
4  Mexican Restaurant  0.12


----M1G----
                       venue  freq
0                Coffee Shop  0.50
1         Mexican Restaurant  0.25
2      Korean BBQ Restaurant  0.25
3  Middle Eastern Restaurant  0.00
4        Monument / Landmark  0.00


----M1H----
                  venue  freq
0                  Bank  0.12
1  Caribbean Restaurant  0.12
2 

In [123]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [124]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = Toronto_grouped['PostalCode']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

PostalCode       1st Most Common Venue  2nd Most Common Venue  \
0        M1B        Fast Food Restaurant    Dumpling Restaurant   
1        M1C  Construction & Landscaping                    Bar   
2        M1E              Breakfast Spot             Restaurant   
3        M1G                 Coffee Shop  Korean BBQ Restaurant   
4        M1H        Caribbean Restaurant       Hakka Restaurant   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0        Discount Store          Distribution Center               Dog Run   
1           Yoga Studio  Eastern European Restaurant               Dog Run   
2     Electronics Store          Rental Car Location        Medical Center   
3    Mexican Restaurant          Dumpling Restaurant   Distribution Center   
4                Bakery                  Gas Station    Athletics & Sports   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop             Drugstore   
1      Doner Restaurant            Donut Shop             Drugstore   
2          Intersection                  Bank    Mexican Restaurant   
3               Dog Run      Doner Restaurant            Donut Shop   
4                  Bank       Thai Restaurant   Fried Chicken Joint   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant          Women's Store  
1          Dumpling Restaurant      Electronics Store  
2                  Yoga Studio                Dog Run  
3                    Drugstore            Yoga Studio  
4             Doner Restaurant    Distribution Center

In [126]:
neighborhoods_venues_sorted.shape

(100, 11)

### Clustering

In [127]:
#  set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('PostalCode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
print(kmeans.labels_.shape)
kmeans.labels_[0:10] 


(100,)


array([1, 4, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

**Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.**

Note that the dataframe Toronto_data have 103 rows and the dataframe neighborhoods_venues_sorted have only 100 rows. Which means there are 3 neighborhood which didn't return any venues (if the venues we removed earlier wasnt the only venues in that neighborhood)

In [129]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_data

# # merge Toronto_grouped with Toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('PostalCode'), on='PostalCode')

Toronto_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

         lat        lon  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             3.0                  Park   
1  43.725882 -79.315572             0.0          Intersection   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio             Drugstore   
1           Pizza Place     French Restaurant          Hockey Arena   
2                Bakery                   Pub                  Park   
3     Accessories Store              Boutique             Gift Shop   
4           Yoga Studio      Sushi Restaurant                   Bar   

    5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0          Discount Store    Distribution Center               Dog Run   
1             Coffee Shop  Portuguese Restaurant            Donut Shop   
2          Breakfast Spot                   Café               Theater   
3  Furniture / Home Store            Event Space           Coffee Shop   
4                Beer Bar          Smoothie Shop        Sandwich Place   

  8th Most Common Venue  9th Most Common Venue       10th Most Common Venue  
0      Doner Restaurant             Donut Shop  Eastern European Restaurant  
1                 Diner         Discount Store          Distribution Center  
2            Beer Store                Brewery                   Shoe Store  
3         Women's Store  Vietnamese Restaurant              Airport Service  
4            Restaurant                   Café           Chinese Restaurant

In [104]:
Toronto_merged.describe()

lat         lon  Cluster Labels
count  103.000000  103.000000      100.000000
mean    43.706291  -79.394423        0.970000
std      0.053049    0.094865        0.481055
min     43.602414  -79.594054        0.000000
25%     43.661955  -79.460544        1.000000
50%     43.704324  -79.388790        1.000000
75%     43.747989  -79.339515        1.000000
max     43.836125  -79.160497        4.000000

As we expected there are 3 rows with no clusters. We have to remove those rows.

In [130]:
Toronto_merged.dropna(inplace=True)

### visualize the resulting clusters

In [132]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['lat'], Toronto_merged['lon'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine CLusters
Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster

#### Cluster 1

In [137]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[2] + list(range(5, Toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
1                                    Victoria Village             0.0   
8                     Parkview Hill, Woodbine Gardens             0.0   
10                                          Glencairn             0.0   
50                                      Humber Summit             0.0   
70                                          Westmount             0.0   
72                        Willowdale, Willowdale West             0.0   
77  Kingsview Village, St. Phillips, Martin Grove ...             0.0   
82            Clarks Corners, Tam O'Shanter, Sullivan             0.0   
89  South Steeles, Silverstone, Humbergate, Jamest...             0.0   
93                             Alderwood, Long Branch             0.0   

   1st Most Common Venue   2nd Most Common Venue 3rd Most Common Venue  \
1           Intersection             Pizza Place     French Restaurant   
8            Pizza Place                Pharmacy             Gastropub   
10                  Park             Pizza Place                Bakery   
50           Pizza Place  Furniture / Home Store            Donut Shop   
70           Pizza Place             Coffee Shop        Discount Store   
72           Pizza Place           Grocery Store           Coffee Shop   
77           Pizza Place                    Park        Sandwich Place   
82           Pizza Place                Pharmacy    Italian Restaurant   
89           Pizza Place           Grocery Store              Pharmacy   
93           Pizza Place                Pharmacy                   Gym   

   4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
1           Hockey Arena           Coffee Shop  Portuguese Restaurant   
8   Gym / Fitness Center             Pet Store           Intersection   
10   Japanese Restaurant                   Pub             Donut Shop   
50                 Diner        Discount Store    Distribution Center   
70        Sandwich Place    Chinese Restaurant           Intersection   
72               Butcher              Pharmacy                    Gym   
77              Bus Line            Donut Shop         Discount Store   
82                  Bank   Fried Chicken Joint     Chinese Restaurant   
89   Fried Chicken Joint        Sandwich Place   Fast Food Restaurant   
93        Sandwich Place           Coffee Shop                    Pub   

   7th Most Common Venue        8th Most Common Venue  \
1             Donut Shop                        Diner   
8         Breakfast Spot           Athletics & Sports   
10        Discount Store          Distribution Center   
50               Dog Run             Doner Restaurant   
70   Dumpling Restaurant  Eastern European Restaurant   
72  Ethiopian Restaurant            Electronics Store   
77   Distribution Center                      Dog Run   
82       Thai Restaurant                  Gas Station   
89            Beer Store                 Liquor Store   
93               Dog Run           Dim Sum Restaurant   

          9th Most Common Venue 10th Most Common Venue  
1                Discount Store    Distribution Center  
8                      Bus Line                   Bank  
10                      Dog Run       Doner Restaurant  
50                    Drugstore         Farmers Market  
70                    Drugstore           Dessert Shop  
72  Eastern European Restaurant    Dumpling Restaurant  
77             Doner Restaurant              Drugstore  
82                 Intersection           Noodle House  
89       Gluten-free Restaurant       Department Store  
93                        Diner         Discount Store

#### Cluster 2

In [138]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[2] + list(range(5, Toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
2                            Regent Park, Harbourfront             1.0   
3                     Lawrence Manor, Lawrence Heights             1.0   
4          Queen's Park, Ontario Provincial Government             1.0   
6                                       Malvern, Rouge             1.0   
7                                            Don Mills             1.0   
9                             Garden District, Ryerson             1.0   
11   West Deane Park, Princess Gardens, Martin Grov...             1.0   
13                                           Don Mills             1.0   
14                                    Woodbine Heights             1.0   
15                                      St. James Town             1.0   
16                                  Humewood-Cedarvale             1.0   
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...             1.0   
18                   Guildwood, Morningside, West Hill             1.0   
19                                         The Beaches             1.0   
20                                         Berczy Park             1.0   
22                                              Woburn             1.0   
23                                             Leaside             1.0   
24                                  Central Bay Street             1.0   
25                                            Christie             1.0   
26                                           Cedarbrae             1.0   
27                                   Hillcrest Village             1.0   
28     Bathurst Manor, Wilson Heights, Downsview North             1.0   
29                                    Thorncliffe Park             1.0   
30                            Richmond, Adelaide, King             1.0   
31                        Dufferin, Dovercourt Village             1.0   
32                                 Scarborough Village             1.0   
33                        Fairview, Henry Farm, Oriole             1.0   
34                     Northwood Park, York University             1.0   
36   Harbourfront East, Union Station, Toronto Islands             1.0   
37                            Little Portugal, Trinity             1.0   
38         Kennedy Park, Ionview, East Birchmount Park             1.0   
39                                     Bayview Village             1.0   
41                        The Danforth West, Riverdale             1.0   
42            Toronto Dominion Centre, Design Exchange             1.0   
43        Brockton, Parkdale Village, Exhibition Place             1.0   
44                     Golden Mile, Clairlea, Oakridge             1.0   
46                                           Downsview             1.0   
47                      India Bazaar, The Beaches West             1.0   
48                      Commerce Court, Victoria Hotel             1.0   
51     Cliffside, Cliffcrest, Scarborough Village West             1.0   
53                                           Downsview             1.0   
54                                     Studio District             1.0   
55                   Bedford Park, Lawrence Manor East             1.0   
56    Del Ray, Mount Dennis, Keelsdale and Silverthorn             1.0   
58                         Birch Cliff, Cliffside West             1.0   
59                         Willowdale, Willowdale East             1.0   
60                                           Downsview             1.0   
62                                            Roselawn             1.0   
63                       Runnymede, The Junction North             1.0   
65   Dorset Park, Wexford Heights, Scarborough Town...             1.0   
67                                    Davisville North             1.0   
68     Forest Hill North & West, Forest Hill Road Park             1.0   
69                       High Park, The Junction South             1.0   
71                                   Wexford, 

#### Cluster 3

In [144]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[2] + list(range(5, Toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
45  York Mills, Silver Hills             2.0   Martial Arts School   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
45           Yoga Studio   Dumpling Restaurant        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
45   Distribution Center               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
45            Donut Shop             Drugstore  Eastern European Restaurant

#### Cluster 4

In [140]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[2] + list(range(5, Toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                                           Parkwoods             3.0   
21                                Caledonia-Fairbanks             3.0   
35      East Toronto, Broadview North (Old East York)             3.0   
40                                          Downsview             3.0   
49           North Park, Maple Leaf Park, Upwood Park             3.0   
61                                      Lawrence Park             3.0   
64                                             Weston             3.0   
66                                    York Mills West             3.0   
85  Milliken, Agincourt North, Steeles East, L'Amo...             3.0   
91                                           Rosedale             3.0   

   1st Most Common Venue       2nd Most Common Venue 3rd Most Common Venue  \
0                   Park           Food & Drink Shop           Yoga Studio   
21                  Park               Women's Store                  Pool   
35                  Park           Convenience Store          Intersection   
40               Airport                        Park           Yoga Studio   
49                  Park  Construction & Landscaping                Bakery   
61                  Park                 Swim School              Bus Line   
64                  Park                 Yoga Studio   Dumpling Restaurant   
66                  Park           Convenience Store           Yoga Studio   
85          Intersection                        Park            Playground   
91                  Park                  Playground                 Trail   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0              Drugstore        Discount Store   Distribution Center   
21             Drugstore                 Diner        Discount Store   
35           Yoga Studio             Drugstore        Discount Store   
40   Dumpling Restaurant        Discount Store   Distribution Center   
49           Yoga Studio   Dumpling Restaurant   Distribution Center   
61           Yoga Studio             Drugstore   Distribution Center   
64        Discount Store   Distribution Center               Dog Run   
66   Dumpling Restaurant        Discount Store   Distribution Center   
85                Bakery            Donut Shop                 Diner   
91           Yoga Studio             Drugstore                 Diner   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                Dog Run      Doner Restaurant            Donut Shop   
21   Distribution Center               Dog Run      Doner Restaurant   
35   Distribution Center               Dog Run      Doner Restaurant   
40               Dog Run      Doner Restaurant            Donut Shop   
49               Dog Run      Doner Restaurant            Donut Shop   
61               Dog Run      Doner Restaurant            Donut Shop   
64      Doner Restaurant            Donut Shop             Drugstore   
66               Dog Run      Doner Restaurant            Donut Shop   
85        Discount Store   Distribution Center               Dog Run   
91        Discount Store   Distribution Center               Dog Run   

         10th Most Common Venue  
0   Eastern European Restaurant  
21                   Donut Shop  
35                   Donut Shop  
40                    Drugstore  
49                    Drugstore  
61          Dumpling Restaurant  
64  Eastern European Restaurant  
66                    Drugstore  
85             Doner Restaurant  
91             Doner Restaurant

#### Cluster 5

In [141]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[2] + list(range(5, Toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
12              Rouge Hill, Port Union, Highland Creek             4.0   
57                                    Humberlea, Emery             4.0   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...             4.0   

          1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
12   Construction & Landscaping                   Bar           Yoga Studio   
57               Baseball Field           Yoga Studio   Distribution Center   
101  Construction & Landscaping        Baseball Field           Yoga Studio   

           4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
12   Eastern European Restaurant               Dog Run      Doner Restaurant   
57                       Dog Run      Doner Restaurant            Donut Shop   
101  Eastern European Restaurant               Dog Run      Doner Restaurant   

    7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
12             Donut Shop             Drugstore          Dumpling Restaurant   
57              Drugstore   Dumpling Restaurant  Eastern European Restaurant   
101            Donut Shop             Drugstore          Dumpling Restaurant   

    10th Most Common Venue  
12       Electronics Store  
57                   Field  
101      Electronics Store